server/src/main/java/io/crate/statistics/TransportAnalyzeAction.java

```diff
@@ -84,26 +87,40 @@ public final class TransportAnalyzeAction {
     private final TransportService transportService;
     private final Schemas schemas;
     private final ClusterService clusterService;
+    private final AtomicReference<CompletableFuture<Samples>> runningAnalyze = new AtomicReference<>();
+    private final Executor executor;
 
     @Inject
     public TransportAnalyzeAction(TransportService transportService,
                                   ReservoirSampler reservoirSampler,
                                   Schemas schemas,
                                   ClusterService clusterService,
-                                  TableStats tableStats) {
+                                  TableStats tableStats,
+                                  ThreadPool threadPool) {
         this.transportService = transportService;
         this.schemas = schemas;
         this.clusterService = clusterService;
+        this.executor = threadPool.executor(ThreadPool.Names.SEARCH);
+
         transportService.registerRequestHandler(
             FETCH_SAMPLES,
-            ThreadPool.Names.SEARCH,
+            ThreadPool.Names.SAME,
             FetchSampleRequest::new,
             // Explicit generic is required for eclipse JDT, otherwise it won't compile
             new NodeActionRequestHandler<FetchSampleRequest, FetchSampleResponse>(
-                req -> completedFuture(new FetchSampleResponse(
-                    reservoirSampler.getSamples(req.relation(), req.columns(), req.maxSamples())))
+                req -> {
+                    runningAnalyze.compareAndExchange(
+                        null,
+                        supplyAsync(
+                            () -> reservoirSampler.getSamples(req.relation(), req.columns(), req.maxSamples()),
+                            executor
+                        )
+                    );
+                    return runningAnalyze.get().thenApply(FetchSampleResponse::new).whenComplete((res, err) -> runningAnalyze.set(null));
+                }
             )
```

Comment:
This is still not really working. This would call `supplyAsync` even if the comparison to `null` fails - always triggering a new `getSamples` run.

It needs something like this:

```diff
diff --git a/server/src/main/java/io/crate/statistics/TransportAnalyzeAction.java b/server/src/main/java/io/crate/statistics/TransportAnalyzeAction.java
index 9d6ed8094c1..c929d6322d6 100644
--- a/server/src/main/java/io/crate/statistics/TransportAnalyzeAction.java
+++ b/server/src/main/java/io/crate/statistics/TransportAnalyzeAction.java
@@ -107,18 +107,24 @@ public TransportAnalyzeAction(TransportService transportService,
             ThreadPool.Names.SAME,
             FetchSampleRequest::new,
             // Explicit generic is required for eclipse JDT, otherwise it won't compile
             new NodeActionRequestHandler<FetchSampleRequest, FetchSampleResponse>(
                 req -> {
-                    runningAnalyze.compareAndExchange(
-                        null,
-                        supplyAsync(
+                    CompletableFuture<Samples> newSamples = new CompletableFuture<>();
+                    CompletableFuture<Samples> previous = runningAnalyze.compareAndExchange(null, newSamples);
+                    if (previous == null) {
+                        newSamples.completeAsync(
                             () -> reservoirSampler.getSamples(req.relation(), req.columns(), req.maxSamples()),
                             executor
-                        )
-                    );
-                    return runningAnalyze.get().thenApply(FetchSampleResponse::new).whenComplete((res, err) -> runningAnalyze.set(null));
+                        );
+                        return newSamples
+                            .thenApply(FetchSampleResponse::new)
+                            .whenComplete((res, err) -> runningAnalyze.set(null));
+                    } else {
+                        return previous
+                            .thenApply(FetchSampleResponse::new);
+                    }
                 }
             )
         );

         transportService.registerRequestHandler(
```

Reply:
>would call supplyAsync

🤦 exactly, thx for catching.